File: clients/rn/src/engine/discovery/InferenceDiscovery.ts
/**
 * Discovers inference endpoints and Wyrdsekai servers on the local network and on-device.
 *
 * Strategy:
 * 1. If a saved URL exists (user explicitly configured), probe it first
 * 2. If household IP known (from SavedHouseholdConfig), probe well-known ports
 * 3. Try common localhost ports (for on-device servers)
 * 4. Return all responsive endpoints
 *
 * Uses short timeouts (2s) so discovery completes quickly even when
 * endpoints are unreachable.
 *
 */

export interface DiscoveredInference {
  /** Base URL of the inference server (e.g., "http://198.51.100.10:11434") */
  url: string;
  /** Server type: "ollama", "llama-server", "openai-compat", or "wyrdsekai" */
  type: 'ollama' | 'llama-server' | 'openai-compat' | 'wyrdsekai';
  /** Human-readable label for display in settings UI */
  label: string;
  /** NATS URL from /health response (Wyrdsekai servers only) */
  natsUrl?: string | null;
  /** Relay URL from /health response (Wyrdsekai servers only) */
  relayUrl?: string | null;
}

const PROBE_TIMEOUT_MS = 2_000;

/**
 * Probe a URL with a short timeout. Returns true if 2xx response.
 */
async function probe(url: string): Promise<boolean> {
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), PROBE_TIMEOUT_MS);
    const response = await fetch(url, { signal: controller.signal });
    clearTimeout(timeout);
    return response.ok;
  } catch {
    return false;
  }
}

/**
 * Probe a Wyrdsekai server's /health endpoint and parse natsUrl/relayUrl.
 * Returns a DiscoveredInference if the server is responsive, null otherwise.
 */
async function probeWyrdsekai(baseUrl: string, label?: string): Promise<DiscoveredInference | null> {
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), PROBE_TIMEOUT_MS);
    const response = await fetch(`${baseUrl}/health`, { signal: controller.signal });
    clearTimeout(timeout);
    if (!response.ok) return null;

    let name = extractHostname(baseUrl);
    let natsUrl: string | null = null;
    let relayUrl: string | null = null;
    try {
      const body = await response.json();
      name = body.name ?? body.server ?? extractHostname(baseUrl);
      natsUrl = body.natsUrl ?? null;
      relayUrl = body.relayUrl ?? null;
    } catch {
      // Parse failure is non-fatal
    }

    return {
      url: baseUrl,
      type: 'wyrdsekai',
      label: label ?? `${name} (${baseUrl})`,
      natsUrl,
      relayUrl,
    };
  } catch {
    return null;
  }
}

function extractHostname(url: string): string {
  try {
    return url.replace(/^https?:\/\//, '').split(':')[0].split('/')[0];
  } catch {
    return url;
  }
}

/**
 * Detect server type from URL. Port 11434 = Ollama, else OpenAI-compatible.
 */
function detectType(url: string): 'ollama' | 'llama-server' | 'openai-compat' {
  return url.includes(':11434') ? 'ollama' : 'openai-compat';
}

/**
 * Discover inference endpoints.
 *
 * @param opts.householdHost IP or hostname of the household server (from mDNS or saved config)
 * @param opts.savedUrl User-configured inference URL (from AsyncStorage)
 * @returns All responsive endpoints, ordered by priority (saved > household > local)
 */
export async function discoverInference(opts?: {
  householdHost?: string;
  savedUrl?: string;
}): Promise<DiscoveredInference[]> {
  const results: DiscoveredInference[] = [];
  const householdHost = opts?.householdHost;
  const savedUrl = opts?.savedUrl;

  // Saved URL first (user explicitly configured)
  if (savedUrl) {
    if (await probe(savedUrl)) {
      results.push({ url: savedUrl, type: detectType(savedUrl), label: 'Saved endpoint' });
    }
  }

  // Probe household host at well-known ports
  if (householdHost) {
    // Wyrdsekai server on port 7070 (primary — gives us natsUrl)
    const wyrdUrl = `http://${householdHost}:7070`;
    const wyrdServer = await probeWyrdsekai(wyrdUrl, `Household server (${householdHost})`);
    if (wyrdServer) {
      results.push(wyrdServer);
    }

    const ollamaUrl = `http://${householdHost}:11434`;
    if (await probe(`${ollamaUrl}/api/tags`)) {
      results.push({ url: ollamaUrl, type: 'ollama', label: `Household Ollama (${householdHost})` });
    }

    const llamaUrl = `http://${householdHost}:8080`;
    if (await probe(`${llamaUrl}/health`)) {
      results.push({ url: llamaUrl, type: 'llama-server', label: `Household llama-server (${householdHost})` });
    }

    // SGLang / vLLM common port
    const sglangUrl = `http://${householdHost}:30000`;
    if (await probe(`${sglangUrl}/health`)) {
      results.push({ url: sglangUrl, type: 'openai-compat', label: `Household inference (${householdHost}:30000)` });
    }
  }

  // Localhost probes (on-device)
  for (const port of [8080, 11434]) {
    const url = `http://localhost:${port}`;
    const probeUrl = port === 11434 ? `${url}/api/tags` : `${url}/health`;
    if (await probe(probeUrl)) {
      const type: DiscoveredInference['type'] = port === 11434 ? 'ollama' : 'llama-server';
      results.push({ url, type, label: `Local (${port})` });
    }
  }

  return results;
}

/**
 * Scan the local /24 subnet for Wyrdsekai servers on port 7070.
 *
 * This is the RN equivalent of KMP's InferenceDiscovery.discover() subnet scan.
 * Probes all 254 IPs in parallel with short timeouts.
 *
 * @param localSubnet Subnet prefix (e.g., "192.168.1"). If not provided, tries common subnets.
 * @returns All discovered Wyrdsekai servers with natsUrl from /health
 */
export async function discoverWyrdsekaiServers(
  localSubnet?: string,
): Promise<DiscoveredInference[]> {
  const subnets = localSubnet ? [localSubnet] : detectLocalSubnets();
  const results: DiscoveredInference[] = [];

  // Probe all IPs in each subnet in parallel
  const probePromises: Promise<DiscoveredInference | null>[] = [];
  for (const subnet of subnets) {
    for (let host = 1; host <= 254; host++) {
      const ip = `${subnet}.${host}`;
      const url = `http://${ip}:7070`;
      probePromises.push(probeWyrdsekai(url));
    }
  }

  const probeResults = await Promise.all(probePromises);
  for (const server of probeResults) {
    if (server && !results.some(r => r.url === server.url)) {
      results.push(server);
    }
  }

  return results;
}

/**
 * Try to detect local subnets. Falls back to common home subnets.
 */
function detectLocalSubnets(): string[] {
  // React Native doesn't have NetworkInterface access.
  // Use common home subnet prefixes as fallback.
  return ['192.168.1', '192.168.10', '192.168.0'];
}

/**
 * Pick the best endpoint from discovered list.
 *
 * Priority: saved > wyrdsekai server > household ollama > household other > local.
 */
export function bestEndpoint(discovered: DiscoveredInference[]): DiscoveredInference | null {
  return (
    discovered.find(d => d.label.startsWith('Saved')) ??
    discovered.find(d => d.type === 'wyrdsekai') ??
    discovered.find(d => d.type === 'ollama' && !d.label.startsWith('Local')) ??
    discovered.find(d => !d.label.startsWith('Local')) ??
    discovered[0] ??
    null
  );
}

File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/discovery/InferenceDiscovery.kt
package org.wyrdsekai.app.engine.discovery

import io.ktor.client.*
import io.ktor.client.plugins.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import kotlinx.coroutines.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.wyrdsekai.app.platform.localIpv4Addresses

/**
 * A discovered Wyrdsekai server on the network.
 *
 * @param url Base URL of the server (e.g., "http://198.51.100.39:7070")
 * @param name Server name or hostname
 * @param label Human-readable label for display
 * @param natsUrl NATS URL from /health (e.g., "nats://198.51.100.39:4222"), null if not present
 * @param relayUrl Relay URL from /health (e.g., "wss://relay.wyrdsekai.org:9222"), null if not present
 */
data class DiscoveredServer(
    val url: String,
    val name: String,
    val label: String,
    val natsUrl: String? = null,
    val relayUrl: String? = null,
    /** Inference config from server — what companion inference is available. */
    val inferenceConfig: InferenceCapability? = null,
)

/**
 * Inference capability advertised by a household server.
 * Parsed from /health response "inference" field.
 */
data class InferenceCapability(
    val available: Boolean = false,
    val provider: String? = null,   // "ollama", "llamacpp", "cloud"
    val baseUrl: String? = null,
    val models: List<String> = emptyList(),
    val companionModel: String? = null,
)

// Keep old type as alias for backward compatibility
typealias DiscoveredInference = DiscoveredServer

/**
 * Discovers Wyrdsekai servers on the local network.
 *
 * SECURITY: Only discovers Wyrdsekai servers (port 7070, /health endpoint).
 * Never probes for raw inference endpoints (Ollama, llama-server, etc.).
 * The phone talks to Wyrdsekai servers only — the server handles inference routing.
 *
 * Discovery strategy:
 * 1. Saved server URL (user explicitly configured)
 * 2. Subnet scan — probe port 7070 on all IPs in the local /24 subnet
 * 3. Return all responsive Wyrdsekai servers
 */
object InferenceDiscovery {

    private const val PROBE_TIMEOUT_MS = 1_500L
    private const val WYRDSEKAI_PORT = 7070
    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Discover Wyrdsekai servers on the network.
     *
     * @param savedUrl User-configured server URL (from TokenStore)
     * @param localSubnet Local subnet prefix (e.g., "192.168.10") — if null, tries to detect
     * @return All responsive Wyrdsekai servers
     */
    suspend fun discover(
        savedUrl: String? = null,
        localSubnet: String? = null,
    ): List<DiscoveredServer> = coroutineScope {
        val results = mutableListOf<DiscoveredServer>()

        // 1. Saved URL first (user explicitly configured — trusted)
        if (!savedUrl.isNullOrBlank()) {
            val server = probeWyrdsekai(savedUrl)
            if (server != null) {
                results.add(server.copy(label = "Saved: ${server.name}"))
            }
        }

        // 2. Subnet scan — probe port 7070 on local network
        val subnets = mutableListOf<String>()
        if (localSubnet != null) {
            subnets.add(localSubnet)
        } else {
            // Common home subnets — try both
            subnets.addAll(detectLocalSubnets())
        }

        // Probe all IPs in parallel (254 per subnet, 1.5s timeout each)
        val probeJobs = subnets.flatMap { subnet ->
            (1..254).map { host ->
                async {
                    val ip = "$subnet.$host"
                    val url = "http://$ip:$WYRDSEKAI_PORT"
                    probeWyrdsekai(url)
                }
            }
        }

        // Collect results
        for (job in probeJobs) {
            val server = job.await()
            if (server != null && results.none { it.url == server.url }) {
                results.add(server)
            }
        }

        results
    }

    /**
     * Pick the best server from discovered list.
     * Priority: saved > first non-local.
     */
    fun bestEndpoint(discovered: List<DiscoveredServer>): DiscoveredServer? {
        return discovered.firstOrNull { it.label.startsWith("Saved") }
            ?: discovered.firstOrNull()
    }

    /**
     * Probe a URL to check if it's a running Wyrdsekai server.
     * Checks GET /health — expects a JSON response with server info.
     * Returns null if not a Wyrdsekai server or unreachable.
     */
    internal suspend fun probeWyrdsekai(baseUrl: String): DiscoveredServer? {
        return try {
            val client = HttpClient {
                install(HttpTimeout) {
                    requestTimeoutMillis = PROBE_TIMEOUT_MS
                    connectTimeoutMillis = PROBE_TIMEOUT_MS
                }
            }
            val response: HttpResponse = client.get("$baseUrl/health")
            client.close()

            if (response.status.value !in 200..299) return null

            // Try to parse server name, natsUrl, relayUrl from health response
            val body = response.bodyAsText()
            var name = extractHostname(baseUrl)
            var natsUrl: String? = null
            var relayUrl: String? = null
            var inferenceConfig: InferenceCapability? = null
            try {
                val obj = json.parseToJsonElement(body).jsonObject
                name = obj["name"]?.jsonPrimitive?.content
                    ?: obj["server"]?.jsonPrimitive?.content
                    ?: extractHostname(baseUrl)
                natsUrl = obj["natsUrl"]?.jsonPrimitive?.content
                relayUrl = obj["relayUrl"]?.jsonPrimitive?.content

                // Parse inference config if present
                val infObj = obj["inference"]?.jsonObject
                if (infObj != null) {
                    inferenceConfig = InferenceCapability(
                        available = infObj["available"]?.jsonPrimitive?.content?.toBoolean() ?: false,
                        provider = infObj["provider"]?.jsonPrimitive?.content,
                        baseUrl = infObj["baseUrl"]?.jsonPrimitive?.content,
                        models = try {
                            infObj["models"]?.jsonArray?.map { it.jsonPrimitive.content } ?: emptyList()
                        } catch (_: Exception) { emptyList() },
                        companionModel = infObj["companionModel"]?.jsonPrimitive?.content,
                    )
                }
            } catch (_: Exception) {
                // Parse failure is non-fatal — name defaults to hostname
            }

            DiscoveredServer(
                url = baseUrl,
                name = name,
                label = "$name ($baseUrl)",
                natsUrl = natsUrl,
                relayUrl = relayUrl,
                inferenceConfig = inferenceConfig,
            )
        } catch (_: Exception) {
            null
        }
    }

    /**
     * Detect local subnet prefixes from common patterns.
     * Returns prefixes like "192.168.1", "192.168.10".
     */
    private fun detectLocalSubnets(): List<String> {
        return try {
            val addresses = localIpv4Addresses()

            addresses.mapNotNull { ip ->
                val parts = ip.split(".")
                if (parts.size == 4 && parts[0] == "192") {
                    "${parts[0]}.${parts[1]}.${parts[2]}"
                } else null
            }.distinct()
        } catch (_: Exception) {
            // Fallback: common home subnets
            listOf("192.168.1", "192.168.10")
        }
    }

    private fun extractHostname(url: String): String {
        return try {
            val stripped = url.removePrefix("http://").removePrefix("https://")
            stripped.substringBefore(":").substringBefore("/")
        } catch (_: Exception) {
            url
        }
    }
}
